### PR TITLE
BUG: fix SparseSeries reindex by using Series implementation

### LINKED ISSUE #15447: Wrong result of pandas.sparse.series.SparseSeries.loc with indexer of length 1
#### Code Sample, a copy-pastable example if possible

```python
In [3]: idx = pd.MultiIndex.from_tuples([('A', 0), ('A', 1), ('B', 0), ('C', 0), ('C', 1)])

In [4]: orig = pd.Series([1, np.nan, np.nan, 3, np.nan], index=idx)

In [5]: sparse = orig.to_sparse()

In [6]: sparse.loc[['A']]
Out[6]: 
A   NaN
dtype: float64
BlockIndex
Block locations: array([], dtype=int32)
Block lengths: array([], dtype=int32)

```
#### Problem description

It should return as below

#### Expected Output
``` python

In [7]: orig.loc[['A']].to_sparse()
Out[7]: 
A  0    1.0
   1    NaN
dtype: float64
BlockIndex
Block locations: array([0], dtype=int32)
Block lengths: array([1], dtype=int32)
```

#### Output of ``pd.show_versions()``

INSTALLED VERSIONS
------------------
commit: f65a6415f15d438432cc6954ead61b052c5d4d60
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.7.0-1-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.utf8
LOCALE: it_IT.UTF-8

pandas: 0.19.0+473.gf65a641
pytest: 3.0.6
pip: 8.1.2
setuptools: 28.0.0
Cython: 0.23.4
numpy: 1.12.0
scipy: 0.18.1
xarray: None
IPython: 5.1.0.dev
sphinx: 1.4.8
patsy: 0.3.0-dev
dateutil: 2.5.3
pytz: 2015.7
blosc: None
bottleneck: 1.2.0
tables: 3.2.2
numexpr: 2.6.0
feather: None
matplotlib: 2.0.0rc2
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.3
lxml: 3.6.4
bs4: 4.5.1
html5lib: 0.999
httplib2: 0.9.1
apiclient: 1.5.2
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.8
s3fs: None
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #15447
 - [x] tests added / passed
 - [x] passes ``git diff master | flake8 --diff``
 - [x] whatsnew entry

I abunded a bit with the tests for ``.reindex()`` (for ``.loc`` there were already quite a bit).

This is completely independent from the approach in #15448 , which also made sense but on which I will come back in a different PR. In the current PR, I'm just exploiting the fact that sparse ``SparseBlock``s "know what to do", and the ``SparseSeries`` itself has nothing particular to do for reindexing.

### CODE REVIEW

File: pandas/sparse/series.py [hunk 1]
```diff
@@ -570,8 +570,7 @@ def copy(self, deep=True):
         return self._constructor(new_data, sparse_index=self.sp_index,
                                  fill_value=self.fill_value).__finalize__(self)
 
-    def reindex(self, index=None, method=None, copy=True, limit=None,
```
Comment (jreback) [inline pandas/sparse/series.py]: leave the signature as it was
use a shared_doc to construct the doc-strings (see how its done in Series itself).

File: pandas/tests/sparse/test_indexing.py [hunk 2]
```diff
@@ -537,6 +542,28 @@ def test_loc_slice(self):
                                   orig.loc['A':'B'].to_sparse())
         tm.assert_sp_series_equal(sparse.loc[:'B'], orig.loc[:'B'].to_sparse())
 
+    def test_reindex(self):
+        orig = self.orig
```
Comment (jreback) [inline pandas/tests/sparse/test_indexing.py]: add the issue number

File: pandas/tests/sparse/test_indexing.py [hunk 3]
```diff
@@ -537,6 +542,28 @@ def test_loc_slice(self):
...
+        exp = orig.reindex([('A', 0), ('C', 1)]).to_sparse()
+        tm.assert_sp_series_equal(res, exp)
+
+        # On specific level:
+        res = sparse.reindex(['A', 'C', 'B'], level=0)
+        exp = orig.reindex(['A', 'C', 'B'], level=0).to_sparse()
+        tm.assert_sp_series_equal(res, exp)
+
+        # single element list (GH 15447)
+        res = sparse.reindex(['A'], level=0)
```
Comment (jreback) [inline pandas/tests/sparse/test_indexing.py]: can you test ``copy`` and other params (limit)

**Comment by jreback** [comment]
lgtm. just some minor doc-string issues.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
can you test with multiple elements as well on the ``.loc`` indexing (unless those tests cases already exist)

File: pandas/tests/sparse/test_indexing.py (hunk 3 above)
Comment (toobaz) [inline pandas/tests/sparse/test_indexing.py]: Added test for ``copy``, others should be already tested in ``pandas/test/series/missing.py``, e.g. [line 659](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/series/test_missing.py#L659).

**Comment by toobaz** [comment]
``.loc`` with multiple elements is already present ([same test](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/sparse/test_indexing.py#L498)).

**Comment by toobaz** [comment]
(The failed test is apparenty a bug in CircleCI)

**Comment by jreback** [comment]
lgtm. ping on green (I restarted circleci).

**Comment by toobaz** [comment]
Failed again! My code is apparently so bad it crashes the install.

**Comment by jreback** [comment]
yeah something happening, though they appear to be working now.

**Comment by toobaz** [comment]
Don't know whether you restarted again, but it's again/still failed.

**Comment by toobaz** [comment]
ping (still stuck)

**Comment by jreback** [comment]
why don't you push again (you can simply do ``git commit --all --amend -C HEAD`` to generate a new commit. This is the only PR where this is happening.

**Comment by toobaz** [comment]
That didn't help, but rebasing on ``master`` apparently did.

File: pandas/sparse/series.py [hunk 4]
```diff
@@ -570,27 +570,13 @@ def copy(self, deep=True):
...
                 **kwargs):
-        """
-        Conform SparseSeries to new Index
-
-        See Series.reindex docstring for general behavior
 
-        Returns
-        -------
-        reindexed : SparseSeries
-        """
```
Comment (jreback) [inline pandas/sparse/series.py]: so I think we need a check  additional kwargs are not pass (IOW, ``tolerance``, ``level``, ``fill_value``). I am actually not sure what happens if you pass these (prob ignored)?

so need some tests for these (should prob just raise ``ValueError`` as not supported by ``SparseSeries``)

further need to have somethin in the doc-string that show them (for ``Series``) or NOT for ``SparseSeries``.

so you need a substitution that can handle this.

File: pandas/tests/sparse/test_indexing.py (hunk 3 above)
Comment (jreback) [inline pandas/tests/sparse/test_indexing.py]: see above, this is True for Series, but NOT for ``SparseSeries``

File: pandas/sparse/series.py (hunk 4 above)
Comment (toobaz) [inline pandas/sparse/series.py]: We are now using the same code as ``Series``, why shouldn't they work?! Anyway: I have now added tests for ``fill_value`` and ``tolerance``. It seems to me that ``level`` is untested even on normal ``Series``, I can add tests for both in a separate PR.

File: pandas/tests/sparse/test_indexing.py (hunk 3 above)
Comment (toobaz) [inline pandas/tests/sparse/test_indexing.py]: Sorry, I meant [line 640](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/series/test_missing.py#L640)

File: pandas/sparse/series.py (hunk 4 above)
Comment (jreback) [inline pandas/sparse/series.py]: @toobaz I don't know if it will or won't work. sure in theory it will, but w/o tests hard to say.

Comment (toobaz) [inline pandas/sparse/series.py]: ``prob ignored`` is different from ``untested``. Anyway: ready to merge?

Comment (jreback) [inline pandas/sparse/series.py]: @toobaz you changed the API by allowing all of these other keywords. They were simply ignored before, now they are not. So obviously not tested. So they either need to be fully tested or raise NotImplementedError. So not ready to merge.

Comment (toobaz) [inline pandas/sparse/series.py]: So just to be sure I understand: by just removing some obsolete and buggy code, I fix multiple bugs (ignoring a documented argument is a bug too) and leave untested only an argument of which (even in the ``Series`` case) I can't even find a working example, let alone an unambiguous documentation. If you don't even have time to at least be more specific in your replies to my comments... well I don't have time neither.

**Comment by toobaz** [comment]
I'ìll push to my personal ``master`` this and other commits; I will try as long as time allows to often rebase on pandas-dev's ``master``. Feel free to cherry-pick when you are interested.

**Comment by jreback** [comment]
@toobaz that is a pretty poor attitude. fixing something is great. making more bugs is not.

**Comment by jorisvandenbossche** [comment]
Jeff, I think @toobaz has shown good attitude in the past and put effort in a lot of issues/PRs. Please be both a bit more patient with each other.

If I understand the issue here, the thing is that now the kwargs are passed, and before they were not (so as a result, silently ignored). And those kwargs are not (fully) tested with a SparseSeries. 
But at the same time, @jreback you are OK with the approach in this PR of handling things to `super(SparseSeries, self).reindex(..)` ? And this fixes some bugs? 
If that is the case, let's go forward with this approach, but just not yet pass all the kwargs (just keep ignoring them) until someone adds tests for them? 

@toobaz Would you still like to make that small change? (not pass the kwargs, or ony pass those for which you added tests) Then I think we can certainly merge this?


> leave untested only an argument of which (even in the Series case) I can't even find a working example, let alone an unambiguous documentation.

@toobaz about which argument are you talking here exactly?

**Comment by jreback** [comment]
@toobaz I simply wanted you to raise ``NotImplemented`` on the addtl kwargs. Sure tests for those would be nice. Your fix was fine, but expanded the API to a bunch of non-tested (and undocumented) code that *should* work, but needs validation.

**Comment by toobaz** [comment]
Thanks to both for looking again into this.

The only (as far as I can see from the docs) untested argument is ``level``, so if there are other "_addtl kwargs_" I'm missing, please name them. ``level`` is (as far as I can see) untested also for regular ``Series``, and the documentation is unclear (to me at least). It is _not_ ``NotImplemented``, it is not _tested_ (again, to the best of my ``grep``ping), for ``SparseSeries`` exactly as for ``Series``. For ``SparseSeries``, before, it _was_ a bug. Now, being still not tested, it _might possibly be_ buggy. As stated above, I would be happy to take care of it in another PR, so that this one can finally be merged. That said, if in the meanwhile you want me to raise ``NotImplemented`` for _that specific_ argument - even though _before_ it was not implemented, while now we have _no specific reason to think_ it won't work, sure, I can do it.

(Then my problem is clearly more general. In my PRs, I see more and more often 1) multiple subsequent revisions asking for things which were not previously asked 2) comments showing zero attention to my replies/to the code being discussed 3) the assumption that every missing test is my responsibility, even though ironically **none** of my PRs **ever** extended the pandas API, and all those that fix bugs do provide related tests. I'm not asking you two to open a general discussion on this, I'm just explaining the reason why I will not be able to spend 2 weeks on getting each trivial bug fix merged: the fact that the time I spend on pandas is not paid doesn't necessarily imply I like to waste it. That's it, feel free not to reply, see you in the next PR, which I hope will see more a constructive conversation.)

**Comment by jorisvandenbossche** [comment]
> level is (as far as I can see) untested also for regular Series, and the documentation is unclear (to me at least).

Small example in the docs: http://pandas.pydata.org/pandas-docs/stable/advanced.html#advanced-reindexing-and-alignment. I think the *behaviour* of reindex(level=) is rather clear (and I also use it regularly), but agree that the docstring explanation is not .. (will open an issue for that). 
Always difficult to find tests in our gigantic test suite, but: there are a few in `tests/test_multilevel.py` ([here](https://github.com/pandas-dev/pandas/blob/09360d80da730008a6a89f38f3780bb1d55f9e25/pandas/tests/test_multilevel.py#L136) and [here](https://github.com/pandas-dev/pandas/blob/09360d80da730008a6a89f38f3780bb1d55f9e25/pandas/tests/test_multilevel.py#L1926)), further also in tests/indexes/test_multi.py.

> For SparseSeries, before, it was a bug. Now, being still not tested, it might possibly be buggy.

I don't know anything of the sparse implementation, so I cannot assess whether it is likely to work out of the box or not. But in any case, this is enabling those keywords (i.e. introducing new functionality that was previously not possible to do), so best to add tests in such a case to assert it is indeed working. 
But let's do that indeed in another PR (and let's open an issue for that), and let's just not pass the kwarg here (leaving the situation exactly how it was) or raise an error (that's the same to me).

**Comment by jorisvandenbossche** [comment]
> I think the behaviour of reindex(level=) is rather clear

See https://github.com/pandas-dev/pandas/issues/15590 for issue on the docstring for level. But actually, this is not that clear as I thought it was :-) 
I always used it to select a level on the calling object (and therefore I thought the docstring explanation was confusing), but actually it also works on the passed Index .. Will comment about that in #15590

**Comment by toobaz** [comment]
Thanks for the reference to the tests and docs I had missed. Actually, while "exploring" more... I noticed that [my PR _does_ introduce a test](https://github.com/pandas-dev/pandas/pull/15461/files#diff-b44a14f548aa4a78ed3680918b3d0198R572) for ``level``, since before I was asked for it :-) And this happens (and I didn't remember) precisely because I copied the tests from ``Series`` and they just worked. So my ``grep``ping was really bad, but the PR should be good?

**Comment by jreback** [comment]
thanks @toobaz

**Pull request closed without merging**